src/main/java/uk/ac/ebi/eva/submission/service/LsriTokenService.java

```diff
@@ -0,0 +1,104 @@
+
+    public String getLsriUserIdFromToken(String deviceCode, int codeExpirationTimeInSeconds) {
+        String userToken = this.pollForToken(deviceCode, codeExpirationTimeInSeconds);
+        // The only definitive attribute we can expect from querying userInfo is the "sub" attribute
+        // See https://connect2id.com/products/server/docs/api/userinfo#claims
+        String tokenAttribute = "sub";
+        return TokenServiceUtil.getUserId(userToken, this.userInfoUrl, tokenAttribute);
+    }
+
+    private String pollForToken(String deviceCode, int maxPollingTimeInSeconds) {
+        int pollingIntervalInSeconds = 5;
+        long startTime = System.currentTimeMillis();
+
+        while (System.currentTimeMillis() - startTime < maxPollingTimeInSeconds * 1000L) {
+            HttpHeaders headers = new HttpHeaders();
+            headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+            headers.setBasicAuth(this.lsriClientId, this.lsriClientSecret);
+
+            MultiValueMap<String, String> map= new LinkedMultiValueMap<>();
+            map.add("scope", "openid");
+            map.add("device_code", deviceCode);
+            map.add("grant_type", "urn:ietf:params:oauth:grant-type:device_code");
+
+            HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(map, headers);
+            RestTemplate restTemplate = new RestTemplate();
+            ResponseEntity<String> response = null;
+            try {
+                response = restTemplate.postForEntity(this.tokenUrl, request, String.class);
+            }
+            catch(HttpClientErrorException ex) {
+                int statusCode = ex.getRawStatusCode();
+                if (statusCode == HttpStatus.BAD_REQUEST.value()) {
+                    // The user has not yet approved the access request
+                    logger.info("Waiting for user approval...");
+                } else {
+                    // An error occurred
+                    logger.error("An error occurred: " + ex.getResponseBodyAsString());
+                    return null;
+                }
+            }
+
+            if (Objects.nonNull(response) && response.getStatusCode() == HttpStatus.OK) {
+                // The user has approved the access request
+                try {
+                    return new ObjectMapper().readTree(response.getBody()).get("access_token").asText();
+                }
+                catch (JsonProcessingException ex) {
+                    ex.printStackTrace();
+                }
+            }
+
+            try {
+                Thread.sleep(pollingIntervalInSeconds * 1000L);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+
+            if (System.currentTimeMillis() - startTime > maxPollingTimeInSeconds * 1000L) {
+                // Stop polling after the maximum polling time
+                logger.error("Polling timed out!");
+                return null;
```

Comment:
Why is this here since the while will do the same thing just after ?

Reply:
Addressed